### PR TITLE
X address compatability

### PR DESCRIPTION
**Pull request by natenichols**
## High Level Overview of Change
ripple-binary-codec can encode x-Addresses.

## Context of Change
Added compatibility for x-Addresses by processing objects before encoding. X-Addresses are broken down into addresses and tags. The object is then encoded.

## Type of Change
[x] New feature (non-breaking change which adds functionality)

Test Plan
Added new tests for transactions containing X-Addresses

### CODE REVIEW

File: src/types/st-object.ts [hunk 1]
```diff
@@ -1,11 +1,68 @@
...
+const SOURCE_TAG_FIELD_NAME = "SourceTag";
+const DESTINATION_TAG_FIELD_NAME = "DestinationTag";
+
+/**
+ * Break down an X-Address into an account and a tag
+ *
+ * @param field Name of field
+ * @param xAddress X-Address cooresponding to field f
+ */
+function handleXAddress(field: string, xAddress: string): JsonObject {
```
Comment (keefertaylor) [inline src/types/st-object.ts:25]: I think your life will be easier if you make this a static function:

```
function parseXAddress(addressFieldValue: string|undefined, tagValue: string|undefined): { address: string, tag: string} {
  // if is x address
     // validate tagValue is null
     // decode and return tuple

  // return raw inputs
}

File: src/types/st-object.ts [hunk 2]
```diff
@@ -1,11 +1,68 @@
...
+  else if (decoded.tag !== false)
+    throw new Error(`${field} cannot have an associated tag`);
+
+  return decoded.tag
+    ? { [field]: decoded.classicAddress, [tagName]: decoded.tag }
+    : { [field]: decoded.classicAddress };
+}
+
+/**
+ * Validate that two objects don't both have the same tag fields
```
Comment (keefertaylor) [inline src/types/st-object.ts:44]: Why is this needed?

File: src/types/st-object.ts [hunk 3]
```diff
@@ -56,6 +113,19 @@ class STObject extends SerializedType {
     const list: BytesList = new BytesList();
     const bytes: BinarySerializer = new BinarySerializer(list);
 
+    Object.keys(value).forEach((f) => {
```
Comment (keefertaylor) [inline src/types/st-object.ts:116]: What does `f` represent? `field`?

File: src/types/st-object.ts [hunk 4]
```diff
@@ -56,6 +113,19 @@ class STObject extends SerializedType {
...
+      if (
```
Comment (keefertaylor) [inline src/types/st-object.ts:117]: Can you comment this condition - it's not obvious what is happening?

Comment (natenichols) [reply src/types/st-object.ts:117]: `Field[f] !== undefined` : so you don't call `name` of undefined 
`type.name` denotes that this is of type `AccountID` 
`/^X/.test(value[f])` tests to see if the value associated with the field starts w/ X and should be treated an an x address.

File: src/types/st-object.ts (hunk 3 above)
Comment (natenichols) [reply src/types/st-object.ts:116]: Yeah, `f` is the string name of the field. Will go back and pick a new name.

File: src/types/st-object.ts (hunk 2 above)
Comment (natenichols) [reply src/types/st-object.ts:44]: Validating that the JSON object doesn't have both an X-Address and a SourceTag. What I probably should do instead is pass it in as a parameter to the `parseXAddress`  function.

File: src/types/account-id.ts [hunk 5]
```diff
@@ -38,6 +42,10 @@ class AccountID extends Hash160 {
    * @returns an AccountID object
    */
   static fromBase58(value: string): AccountID {
+    if (!isValidClassicAddress(value)) {
```
Comment (intelliot) [inline src/types/account-id.ts:45]: Is this necessary? I think `decodeAccountID` throws an error if the value is not a valid classic address.

File: src/types/hash.ts [hunk 6]
```diff
@@ -9,24 +9,26 @@ class Hash extends Comparable {
 
   constructor(bytes: Buffer) {
     super(bytes);
+    if (this.bytes.byteLength !== (this.constructor as typeof Hash).width) {
+      throw new Error(`Invalid Hash length ${this.bytes.byteLength}`);
```
Comment (intelliot) [inline src/types/hash.ts:13]: Could we add a test showing what triggers this error?

File: src/types/st-object.ts [hunk 7]
```diff
@@ -1,11 +1,53 @@
...
+  xAddressToClassicAddress,
+  isValidXAddress,
+} from "ripple-address-codec";
 import { BinaryParser } from "../serdes/binary-parser";
 import { BinarySerializer, BytesList } from "../serdes/binary-serializer";
 
-const OBJECT_END_MARKER = Buffer.from([0xe1]);
-const OBJECT_END_MARKER_NAME = "ObjectEndMarker";
-const OBJECT_FIELD_TYPE_NAME = "STObject";
+const OBJECT_END_MARKER_BYTES = Buffer.from([0xe1]);
```
Comment (intelliot) [inline src/types/st-object.ts:10]: This looks like one byte, so why is the name `BYTES` (plural)?

```suggestion
const OBJECT_END_MARKER_BYTE = Buffer.from([0xe1]);
```

(If renamed, all uses of this constant would also need to be changed.)

File: src/types/st-object.ts [hunk 8]
```diff
@@ -1,11 +1,53 @@
...
+function handleXAddress(field: string, xAddress: string): JsonObject {
+  const decoded = xAddressToClassicAddress(xAddress);
+
+  let tagName;
+  if (field === DESTINATION) tagName = DEST_TAG;
+  else if (field === ACCOUNT) tagName = SOURCE_TAG;
+  else if (decoded.tag !== false)
+    throw new Error(`${field} cannot have an associated tag`);
+
+  return decoded.tag
```
Comment (intelliot) [inline src/types/st-object.ts:33]: I think there's a rare edge case where a tag of `0` may fail to be used. While looking into this, I spotted a mistake in the ripple-address-codec README, and [opened a PR to fix it](https://github.com/ripple/ripple-address-codec/pull/114).

Checking that the tag is not `false` should fix the edge case:

```suggestion
  return decoded.tag !== false
```

File: test/x-address.test.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    PreviousTxnLgrSeq: 7,
+    LedgerEntryType: 'AccountRoot',
+    PreviousTxnID: 'DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF',
+    Flags: 0,
+    Sequence: 1,
+    Balance: '10000000000',
+    SourceTag: 12345,
+  }
+
+describe("X-Address account is equivilent to an r-address w/ SourceTag", () => {
```
Comment (intelliot) [inline test/x-address.test.js:78]:
```suggestion
describe("X-Address Account is equivalent to a classic address w/ SourceTag", () => {
```

File: test/x-address.test.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    test("Can encode with x-Address", () => {
+        expect(encoded_x).toEqual(encoded_r);
+    })
+
+    test("decoded X-address is object w/ source and tag", () => {
+        let decoded_x = decode(encoded_x);
+        expect(decoded_x).toEqual(json_r1);
+    })
+
+    test("Encoding issuer X-Address w/ null destintion tag", () => {
```
Comment (intelliot) [inline test/x-address.test.js:90]: Consider installing a spell checker to help catch typos like `destintion`. I've used [this spell checker for VS Code](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker).

Technically, the absence of a tag is more like `undefined` than `null`, which is a separate value in JS (though this might be debatable). Still, I think it makes a little more sense to say that the tag is undefined like so:

```suggestion
    test("Encoding issuer X-Address w/ undefined destination tag", () => {
```

File: test/x-address.test.js [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    test("Throws when Account has both X-Addr and Destination Tag", () => {
+        expect(() => encode(invalid_json_x_and_tagged)).toThrow(new Error("Cannot have Account X-Address and SourceTag"));
+    });
+})
+
+describe('ripple-binary-codec x-address test', function () {
+    function makeSuite (name, entries) {
+        describe(name, function () {
+        entries.forEach((t, testN) => {
+            test(`${name}[${testN}] encodes x-address json equivilent to as r-address json`,
```
Comment (intelliot) [inline test/x-address.test.js:109]:
```suggestion
            test(`${name}[${testN}] encodes X-address json equivalent to classic address json`,
```

File: test/x-address.test.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            () => {
+                expect(encode(t.rjson)).toEqual(encode(t.xjson))
+            })
+            test(`${name}[${testN}] decodes encoded x-address json equivilent to as r-address json`, () => {
```
Comment (intelliot) [inline test/x-address.test.js:113]:
```suggestion
            test(`${name}[${testN}] decodes encoded X-address json equivalent to classic address address json`, () => {
```

File: src/types/account-id.ts (hunk 5 above)
Comment (natenichols) [reply src/types/account-id.ts:45]: Okay, if `decodeAccountID` throws then we don't need it.

Comment (natenichols) [reply src/types/account-id.ts:45]: Let me look into it real quick
